File: examples/03-chain/after.py
"""Example 03 — Chain of Responsibility, with the pattern.

One handler per provider, and the fallthrough that returned zero becomes an
explicit fallback that raises. Adding a provider is a new function and one list
entry — the function that already worked is never reopened.

Note the parameter name: `next_`, because `next` is a builtin.

Run it:  PYTHONPATH=packages/py python3 examples/03-chain/after.py
"""

from gof_patterns import chain


def from_stripe(event: dict, next_):
    if event["source"] == "stripe":
        return {"cents": event.get("data", {}).get("total", 0)}
    return next_()


def from_spei(event: dict, next_):
    if event["source"] == "spei":
        return {"cents": round(event.get("amount", 0) * 100)}
    return next_()


# The provider that arrived after the original code was written.
def from_paypal(event: dict, next_):
    if event["source"] == "paypal":
        return {"cents": round(event.get("gross", 0) * 100)}
    return next_()


def unknown(event: dict):
    # The decision that used to hide in a `return {"cents": 0}`, now an argument.
    raise ValueError(f"unknown source: {event['source']}")


normalize = chain([from_stripe, from_spei, from_paypal], fallback=unknown)

print(f"stripe → {normalize({'source': 'stripe', 'data': {'total': 1999}})['cents']} centavos")
print(f"spei   → {normalize({'source': 'spei', 'amount': 19.99})['cents']} centavos")
print(f"paypal → {normalize({'source': 'paypal', 'gross': 19.99})['cents']} centavos")

refused = ""
try:
    normalize({"source": "bitcoin"})
except ValueError as error:
    refused = str(error)
print(f"desconocido → {refused}")

assert normalize({"source": "paypal", "gross": 19.99})["cents"] == 1999
assert refused == "unknown source: bitcoin"
print("\n✅ un proveedor nuevo es una función más, y lo desconocido truena")

File: examples/03-chain/after.ts
/**
 * Example 03 — Chain of Responsibility, with the pattern.
 *
 * One handler per provider, and the `else` that returned zero becomes an explicit
 * fallback that throws. Adding a provider is a new function and one array entry —
 * the function that already worked is never reopened.
 *
 * Run it:  node examples/03-chain/after.ts
 */
import assert from 'node:assert/strict';
import { chain } from '../../packages/ts/src/index.ts';

type Incoming = {
  source: string;
  amount?: number;
  gross?: number;
  data?: { total?: number };
};
type Amount = { cents: number };

const fromStripe = (e: Incoming, next: () => Amount): Amount =>
  e.source === 'stripe' ? { cents: e.data?.total ?? 0 } : next();

const fromSpei = (e: Incoming, next: () => Amount): Amount =>
  e.source === 'spei' ? { cents: Math.round((e.amount ?? 0) * 100) } : next();

// The provider that arrived after the original code was written. A new function,
// and one entry below — nothing else moves.
const fromPaypal = (e: Incoming, next: () => Amount): Amount =>
  e.source === 'paypal' ? { cents: Math.round((e.gross ?? 0) * 100) } : next();

const normalize = chain<Incoming, Amount>([fromStripe, fromSpei, fromPaypal], (e) => {
  // The decision that used to hide in a `return { cents: 0 }`, now an argument.
  throw new Error(`unknown source: ${e.source}`);
});

console.log(`stripe → ${normalize({ source: 'stripe', data: { total: 1999 } }).cents} centavos`);
console.log(`spei   → ${normalize({ source: 'spei', amount: 19.99 }).cents} centavos`);
console.log(`paypal → ${normalize({ source: 'paypal', gross: 19.99 }).cents} centavos`);

let refused = '';
try {
  normalize({ source: 'bitcoin' });
} catch (error) {
  refused = (error as Error).message;
}
console.log(`desconocido → ${refused}`);

assert.equal(normalize({ source: 'paypal', gross: 19.99 }).cents, 1999);
assert.equal(refused, 'unknown source: bitcoin');

// Same helper for async, because Res can be a promise. No separate "async" version.
const lookup = chain<string, Promise<string>>(
  [
    async (id, next) => (id === 'cache' ? 'from-cache' : next()),
    async (id, next) => (id === 'db' ? 'from-db' : next()),
  ],
  async () => 'not-found',
);

assert.equal(await lookup('db'), 'from-db');
assert.equal(await lookup('zzz'), 'not-found');
console.log('\n✅ un proveedor nuevo es una función más, y lo desconocido truena');
